### ISSUE TITLE
ROC with KVM

### ISSUE THREAD

**Issue by insujang** [RadeonOpenCompute/ROCK-Kernel-Driver#26]
Created a new issue from [here](https://github.com/RadeonOpenCompute/ROCK-Kernel-Driver/issues/25), as the topic is slightly different.

I am trying to use ROC on a virtualized guest machine.

My machine is
CPU: Intel Core i7 6700
GPU: AMD Radeon RX 480
which satisfies AMD ROC requirements (PCIe 3.0, AtomicOps), and it actually works on the host.

in [here](https://www.phoronix.com/forums/forum/linux-graphics-x-org-drivers/open-source-amd-linux/911118-radeon-open-compute-1-3-platform-brings-polaris-other-features?p=911126#post911126), ROCm 1.3 officially announced that it supports kvm passthrough-ed virtual machine. However, in my setup, it is impossible due to the following error when I run the sample on the guest machine.

```
kfd kfd: skipped device 1002:67df, PCI rejects atomics
```

Virtualization setup is as follows.
Hypervisor: QEMU 2.7.92 + KVM (kernel 4.11)
QEMU script:
```
sudo vfio-bind 0000:01:00.0 0000:01:00.1 &&

sudo qemu-system-x86_64 -enable-kvm -cpu host,kvm=off \
-m 8192 -smp 8 -M q35 \
-vga none -nographic \
-usb \
-device usb-host,hostbus=1,hostaddr=2 \
-device vfio-pci,host=01:00.0,multifunction=on,x-vga=on \
-device vfio-pci,host=01:00.1 \
-drive if=pflash,format=raw,readonly,file=OVMF_CODE.fd \
-drive if=pflash,format=raw,file=OVMF_VARS.fd \
-hda ubuntu.qcow2 \
-net nic \
-net user,hostfwd=tcp::54321-:22
```

I didn't use ioh3420 virtualized pcie controller as if there is in the system, both guest machine and host machine freeze during loading amdgpu driver.

But AMD says it officially support virtualized environment, could you tell me more detailed setup that I missed?

Thanks!

**Comment by gstoner** [comment]
We talked you about this, you have to get the PCIe Atomics to poke through the KVM,  we had it running in the labs,  the issue right now is the current version of KVM is missing this capability.  We are working with key vendors to address the situation thank you for your patience.  For now you need to run ROCm on the metal

**Comment by gstoner** [comment]
This CPU is not supported by ROCm. I added some more info on why we us PCIe Atomics 

System Requirements
To use ROCm on your system you need the following:

ROCm Capable CPU and GPU
PCIe Gen 3 Enabled CPU with PCIe Platform Atomics
ROCm enabled GPU’s
Radeon Instinct Family MI25, MI8, MI6
Radeon Vega Frontier Edition
Broader Set of Tested Hardware

ROCm is an extension of HSA platform architecture, so it shares the queueing model, memory model, signaling and synchronization protocols. Platform atomics are integral to perform queuing and signaling memory operations where there may be multiple-writers across CPU and GPU agents.

The full list of HSA system architecture platform requirements are here: http://www.hsafoundation.com/html/HSA_Library.htm#SysArch/Topics/01_Overview/list_of_requirements.htm 

For ROCm the Platform atomics are used in ROCm in the following ways:

Update HSA queue’s read_dispatch_id: 64bit atomic add used by the command processor on the GPU agent to update the packet ID it processed.
Update HSA queue’s write_dispatch_id: 64bit atomic add used by the CPU and GPU agent to support multi-writer queue insertions.
Update HSA Signals – 64bit atomic ops are used for CPU & GPU synchronization.
The PCIe Platform Atomics are CAS, FetchADD, SWAP

Here is document on PCIe Atomics https://pcisig.com/sites/default/files/specification_documents/ECN_Atomic_Ops_080417.pdf

In ROCm, we also take advantage of PCIe ID based ordering technology for P2P when the GPU originates two writes to two different targets:  

write to another GPU memory,
then write to system memory to indicate transfer complete.
They are routed off to different ends of the computer but we want to make sure the write to system memory to indicate transfer complete occurs AFTER P2P write to GPU has complete.

**Comment by insujang** [comment]
@gstoner Thank you for the reply, and sorry to make you leave the same answer. I didn't understand what you said 'cpu doesn't support ROCm'. My word 'Intle Core i7 supports ROCm' should confuse you. 
What I meant was that the PCIe controller embedded in this CPU supports PCIe 3.0 and AtomicOps, so it makes an AMD GPU in the system available for ROCm.

What I still don't understand is that, AMD announced ROCm support ***GPU hardware virtualization***  via KVM pass-through since version 1.3. [link](http://www.amd.com/en-us/press-releases/Pages/rocm-platform-2016nov14.aspx)
So does it mean if I want to use ROCm with AMD GPUs in a virtual machine, I also must use a CPU that supports ROCm? That doesn't make sense to me and the reason I didn't understand what you said.

Could you make it more clear?

Let me summarize my understanding.
1. AMD releases ROCm v1.3 and says it supports GPU virtualization via KVM pass-through.
2. You said current KVM is missing the capability. (right now ROCm is version 1.6!)
3. Intel CPU doesn't support ROCm **by itself**. (So it can do a role of CPU agent, I got it.)

I'm confusing that statement 1 and 2 are conflicting each other.

Thanks!!

**Comment by gstoner** [comment]
Sorry your CPU is supported by ROCm someone was asking if Xeon V1 was supported

**Comment by gstoner** [comment]
I will dig into what going on with that statement by marketing. 

i run the engineering team and can tell what really works or not and what we working on. 

We are working on SRIOV support aka MxGPU for ROCm that I can tell you is under development.   

I went back through the emails we did get this to work by enabling atomics using the correct offset in the PCI config space makes ROCm apps execute and complete correctly. See bellow. 

 KVM can get the AtomicOp Routing working, we need the AtomicOp Requester Enable flag to get set on the GPU.  What was happening is PCI configuration space (or update to Device Capabilities 2 register) isn’t writeable from within the guest os to prevent escapes from the VM .  We need the host os/vm to set the config bit.

This what happing when it is wrong Looking at lspci from the host shows:
 
sudo lspci -s 0b:00.0 -xxx
0b:00.0 Display controller: Advanced Micro Devices, Inc. [AMD/ATI] Device 7300 (rev c1)
00: 02 10 00 73 07 05 10 00 c1 00 00 03 10 00 80 00
10: 0c 00 00 e0 ff 2f 00 00 0c 00 00 f0 ff 2f 00 00
20: 01 00 00 00 00 00 b0 c7 00 00 00 00 02 10 35 0b
30: 00 00 fe ff 48 00 00 00 00 00 00 00 0b 01 00 00
40: 00 00 00 00 00 00 00 00 09 50 08 00 02 10 35 0b
50: 01 58 03 f6 08 00 00 00 10 a0 12 00 a1 8f 00 00
60: 2c 28 00 00 03 0d 40 08 00 00 01 11 00 00 00 00
70: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 70 00
80: **00**00 00 00 0e 00 00 00 03 00 01 00 00 00 00 00
90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
a0: 05 00 81 00 98 03 e0 fe 00 00 00 00 00 00 00 00
b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
e0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
f0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 
Setting the bit:
sudo setpci -v -d *:7300 80.b=40
0000:0c:00.0 @80 40
0000:0b:00.0 @80 40
 
Which results in:
sudo lspci -s 0b:00.0 -xxx
0b:00.0 Display controller: Advanced Micro Devices, Inc. [AMD/ATI] Device 7300 (rev c1)
00: 02 10 00 73 07 05 10 00 c1 00 00 03 10 00 80 00
10: 0c 00 00 e0 ff 2f 00 00 0c 00 00 f0 ff 2f 00 00
20: 01 00 00 00 00 00 b0 c7 00 00 00 00 02 10 35 0b
30: 00 00 fe ff 48 00 00 00 00 00 00 00 0b 01 00 00
40: 00 00 00 00 00 00 00 00 09 50 08 00 02 10 35 0b
50: 01 58 03 f6 08 00 00 00 10 a0 12 00 a1 8f 00 00
60: 2c 28 00 00 03 0d 40 08 00 00 01 11 00 00 00 00
70: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 70 00
80: **40**00 00 00 0e 00 00 00 03 00 01 00 00 00 00 00
90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
a0: 05 00 81 00 98 03 e0 fe 00 00 00 00 00 00 00 00
b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
e0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
f0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 
The bit must be set after the VM finishes booting to avoid the ASIC reset clearing it.


 I file  a bug into Redhat for PCI atomic support for KVM pass though

**Comment by insujang** [comment]
@gstoner 
Thank you for your detailed and kind explanation!!
Let me try it and leave a result in here.

Thank you!

**Comment by gstoner** [comment]
I ping my contact on the Xen project and also REHL to see if we can get this fixed.

**Comment by gstoner** [comment]
I am going to also ping VMware contact but that will take a lot longer as you know

**Comment by insujang** [comment]
@gstoner Sorry to late. I reinstalled everything from scratch (host kernel, guest kernel, qemu, ROCm, etc) since packages are malfunctioning.

In short, your comment works like a charm. Thank you very much! 
But still there is an additional work to be needed.

1. amdgpu and amdkfd driver is set to be load during boot procedure. When amdkfd driver performs `kgd2kfd_probe()` (/ROCK/drivers/gpu/drm/amd/amdkfd/kfd_device.c:286), it calls the kernel function `pci_enable_atomic_ops_to_root(pdev)`.

This function looks like as follows.
```
int pci_enable_atomic_ops_to_root(struct pci_dev *dev)
{
        struct pci_bus *bus = dev->bus;

        // several condition checks
        pcie_capability_set_word(dev, PCI_EXP_DEVCTL2, PCI_EXP_DEVCTL2_ATOMIC_REQ);
        return 0;
}
```

Your comment must mean that setting the word PCI_EXP_DEVCTL2 from a guest machine doesn't work, so we need to modify by manually. I did it.

2. However, after reboot, amdkfd driver still reports PCI rejects atomics.
The reason is that, AMDGPU is recognized as a PCI device, not a PCIe device in the guest machine, so `pci_enable_atomic_ops_to_root()` returns non-zero value due to the following condition check.

```
int pci_enable_atomic_ops_to_root(struct pci_dev *dev)
{
        struct pci_bus *bus = dev->bus;
        if (!pci_is_pcie(dev))  // returns 0, which makes the condition true
                return -EINVAL;
       ...
}
```

The reason is that I did not use ioh3420 pcie controller in the guest machine. If I use ioh3420, during loading amdgpu driver, both the guest machine and the host machine freezes, no choice but to press a hard reset button. That's the reason I removed ioh3420 pcie controller. So amdgpu right now is attached directly to bus pcie.0, and recognized as a PCI device.

3. To make it work, I modified `pci_enable_atomic_ops_to_root()` in ROCK kernel to bypass all condition checks as follows.
```
int pci_enable_atomic_ops_to_root(struct pci_dev *dev)
{
        struct pci_bus *bus = dev->bus;
        pcie_capability_set_word(dev, PCI_EXP_DEVCTL2,
                                 PCI_EXP_DEVCTL2_ATOMIC_REQ);

        return 0;
}
```
Well, I understand that `pcie_capability_set_word(dev, PCI_EXP_DEVCTL2, PCI_EXP_DEVCTL2_ATOMIC_REQ);` has no effect according to your comment. But just modified like this.
Then, it should work when amdkfd driver is being probed.

3. I blacklisted amdgpu driver during guest kernel boot. It is necessary that you said the bit must be set after th VM finishes booting, but should also be done before amdgpu is initialized. Before modprobing amdgpu kernel, I set bit 80 of GPU's PCI configuration space to 0x40 and type `sudo modprobe amdgpu`. Now kfd driver adds dGPU to ROC topology as follows.

```
[   98.575698] kfd kfd: Allocated 3969056 bytes on gart for device 1002:67df
[   98.575704] Virtual CRAT table created for GPU
[   98.575705] Parsing CRAT table with 1 nodes
[   98.575709] Creating topology SYSFS entries
[   98.575770] Topology: Add dGPU node [0x67df:0x1002]
[   98.575793] kfd kfd: Reserved 2 pages for cwsr.
[   98.575812] kfd kfd: added device 1002:67df
```
And then, the sample runs successfully.

I don't know why the entire machine freezes if I use ioh3420 controller, but anyway it works now.
Thank you very much!

**Comment by bd4** [comment]
I am seeing this now with Debian bullseye host running 5.10.0-7-amd64, KVM 5.2.0 (Debian 1:5.2+dfsg-10+b2) and guest Ubuntu 20.04 with HWE kernel (5.8 something). This is while passing through an MSI RX 570 ARMOR MK2 8G OC. Surprised to see this is still an issue in some cases. Is there some combinations of versions that is known to work? I also tried using latest rock-dkms package from latest ubuntu ROCm repo, and that resulted in a failed firmware load, the in-kernel 5.8 version was at least loading firmware. I haven't been able to test the set-pci workaround yet - amdgpu is loading even though I blacklisted it, and if I try after doing an rmmod things get messed up (PCI reset bug?) and I have to reboot the host.
